### PR TITLE
Support nested tuples

### LINKED ISSUE #1886: Nested tuples do not work with lists
```python
from lpython import i32, f64

def test_list_pop():
    l1: list[tuple[i32, f64, tuple[f64, i32]]]
    x: tuple[i32, f64, tuple[f64, i32]]

    l1 = [(1, 1.0, (0.0, 0)), (2, 4.0, (3.0, 1)), (3, 6.0, (5.0, 2))]
    x = l1.pop(0)
    print(x)        # incorrect
    l1.append((-1, -2.0, (-3.0, -2)))
    print(x)
    print(l1)       # correct

test_list_pop()
```

**Error**

```llvm
; ModuleID = 'LFortran'
source_filename = "LFortran"

%list = type { i32, i32, %tuple.0* }
%tuple.0 = type { i32, double, %tuple }
%tuple = type { double, i32 }

define void @__module__global_symbols__lpython_main_program() {
.entry:
  call void @__module__global_symbols_test_list_pop()
  br label %return

return:                                           ; preds = %.entry
  ret void
}

define void @__module__global_symbols_test_list_pop() {
.entry:
  %l1 = alloca %list, align 8
  %0 = call i8* (i32, ...) @_lfortran_malloc(i32 32)
  %1 = bitcast i8* %0 to %tuple.0*
  %2 = getelementptr %list, %list* %l1, i32 0, i32 2
  store %tuple.0* %1, %tuple.0** %2, align 8
  %3 = getelementptr %list, %list* %l1, i32 0, i32 0
  store i32 0, i32* %3, align 4
  %4 = getelementptr %list, %list* %l1, i32 0, i32 1
  store i32 1, i32* %4, align 4
  %x = alloca %tuple.0, align 8
  %const_list = alloca %list, align 8
  %5 = call i8* (i32, ...) @_lfortran_malloc(i32 96)
  %6 = bitcast i8* %5 to %tuple.0*
  %7 = getelementptr %list, %list* %const_list, i32 0, i32 2
  store %tuple.0* %6, %tuple.0** %7, align 8
  %8 = getelementptr %list, %list* %const_list, i32 0, i32 0
  store i32 3, i32* %8, align 4
  %9 = getelementptr %list, %list* %const_list, i32 0, i32 1
  store i32 3, i32* %9, align 4
  %const_tuple = alloca %tuple.0, align 8
  %const_tuple1 = alloca %tuple, align 8
  %10 = getelementptr %tuple, %tuple* %const_tuple1, i32 0, i32 0
  store double 0.000000e+00, double* %10, align 8
  %11 = getelementptr %tuple, %tuple* %const_tuple1, i32 0, i32 1
  store i32 0, i32* %11, align 4
  %12 = getelementptr %tuple.0, %tuple.0* %const_tuple, i32 0, i32 0
  store i32 1, i32* %12, align 4
  %13 = getelementptr %tuple.0, %tuple.0* %const_tuple, i32 0, i32 1
  store double 1.000000e+00, double* %13, align 8
  %14 = getelementptr %tuple.0, %tuple.0* %const_tuple, i32 0, i32 2
  store %tuple* %const_tuple1, %tuple* %14, align 8
  %15 = getelementptr %list, %list* %const_list, i32 0, i32 2
  %16 = load %tuple.0*, %tuple.0** %15, align 8
  %17 = getelementptr inbounds %tuple.0, %tuple.0* %16, i32 0
  %18 = getelementptr %tuple.0, %tuple.0* %const_tuple, i32 0, i32 0
  %19 = load i32, i32* %18, align 4
  %20 = getelementptr %tuple.0, %tuple.0* %17, i32 0, i32 0
  store i32 %19, i32* %20, align 4
  %21 = getelementptr %tuple.0, %tuple.0* %const_tuple, i32 0, i32 1
  %22 = load double, double* %21, align 8
  %23 = getelementptr %tuple.0, %tuple.0* %17, i32 0, i32 1
  store double %22, double* %23, align 8
  %24 = getelementptr %tuple.0, %tuple.0* %const_tuple, i32 0, i32 2
  %25 = getelementptr %tuple.0, %tuple.0* %17, i32 0, i32 2
  %26 = getelementptr %tuple, %tuple* %24, i32 0, i32 0
  %27 = load double, double* %26, align 8
  %28 = getelementptr %tuple, %tuple* %25, i32 0, i32 0
  store double %27, double* %28, align 8
  %29 = getelementptr %tuple, %tuple* %24, i32 0, i32 1
  %30 = load i32, i32* %29, align 4
  %31 = getelementptr %tuple, %tuple* %25, i32 0, i32 1
  store i32 %30, i32* %31, align 4
  %const_tuple2 = alloca %tuple.0, align 8
  %const_tuple3 = alloca %tuple, align 8
  %32 = getelementptr %tuple, %tuple* %const_tuple3, i32 0, i32 0
  store double 3.000000e+00, double* %32, align 8
  %33 = getelementptr %tuple, %tuple* %const_tuple3, i32 0, i32 1
  store i32 1, i32* %33, align 4
  %34 = getelementptr %tuple.0, %tuple.0* %const_tuple2, i32 0, i32 0
  store i32 2, i32* %34, align 4
  %35 = getelementptr %tuple.0, %tuple.0* %const_tuple2, i32 0, i32 1
  store double 4.000000e+00, double* %35, align 8
  %36 = getelementptr %tuple.0, %tuple.0* %const_tuple2, i32 0, i32 2
  store %tuple* %const_tuple3, %tuple* %36, align 8
  %37 = getelementptr %list, %list* %const_list, i32 0, i32 2
  %38 = load %tuple.0*, %tuple.0** %37, align 8
  %39 = getelementptr inbounds %tuple.0, %tuple.0* %38, i32 1
  %40 = getelementptr %tuple.0, %tuple.0* %const_tuple2, i32 0, i32 0
  %41 = load i32, i32* %40, align 4
  %42 = getelementptr %tuple.0, %tuple.0* %39, i32 0, i32 0
  store i32 %41, i32* %42, align 4
  %43 = getelementptr %tuple.0, %tuple.0* %const_tuple2, i32 0, i32 1
  %44 = load double, double* %43, align 8
  %45 = getelementptr %tuple.0, %tuple.0* %39, i32 0, i32 1
  store double %44, double* %45, align 8
  %46 = getelementptr %tuple.0, %tuple.0* %const_tuple2, i32 0, i32 2
  %47 = getelementptr %tuple.0, %tuple.0* %39, i32 0, i32 2
  %48 = getelementptr %tuple, %tuple* %46, i32 0, i32 0
  %49 = load double, double* %48, align 8
  %50 = getelementptr %tuple, %tuple* %47, i32 0, i32 0
  store double %49, double* %50, align 8
  %51 = getelementptr %tuple, %tuple* %46, i32 0, i32 1
  %52 = load i32, i32* %51, align 4
  %53 = getelementptr %tuple, %tuple* %47, i32 0, i32 1
  store i32 %52, i32* %53, align 4
  %const_tuple4 = alloca %tuple.0, align 8
  %const_tuple5 = alloca %tuple, align 8
  %54 = getelementptr %tuple, %tuple* %const_tuple5, i32 0, i32 0
  store double 5.000000e+00, double* %54, align 8
  %55 = getelementptr %tuple, %tuple* %const_tuple5, i32 0, i32 1
  store i32 2, i32* %55, align 4
  %56 = getelementptr %tuple.0, %tuple.0* %const_tuple4, i32 0, i32 0
  store i32 3, i32* %56, align 4
  %57 = getelementptr %tuple.0, %tuple.0* %const_tuple4, i32 0, i32 1
  store double 6.000000e+00, double* %57, align 8
  %58 = getelementptr %tuple.0, %tuple.0* %const_tuple4, i32 0, i32 2
  store %tuple* %const_tuple5, %tuple* %58, align 8
  %59 = getelementptr %list, %list* %const_list, i32 0, i32 2
  %60 = load %tuple.0*, %tuple.0** %59, align 8
  %61 = getelementptr inbounds %tuple.0, %tuple.0* %60, i32 2
  %62 = getelementptr %tuple.0, %tuple.0* %const_tuple4, i32 0, i32 0
  %63 = load i32, i32* %62, align 4
  %64 = getelementptr %tuple.0, %tuple.0* %61, i32 0, i32 0
  store i32 %63, i32* %64, align 4
  %65 = getelementptr %tuple.0, %tuple.0* %const_tuple4, i32 0, i32 1
  %66 = load double, double* %65, align 8
  %67 = getelementptr %tuple.0, %tuple.0* %61, i32 0, i32 1
  store double %66, double* %67, align 8
  %68 = getelementptr %tuple.0, %tuple.0* %const_tuple4, i32 0, i32 2
  %69 = getelementptr %tuple.0, %tuple.0* %61, i32 0, i32 2
  %70 = getelementptr %tuple, %tuple* %68, i32 0, i32 0
  %71 = load double, double* %70, align 8
  %72 = getelementptr %tuple, %tuple* %69, i32 0, i32 0
  store double %71, double* %72, align 8
  %73 = getelementptr %tuple, %tuple* %68, i32 0, i32 1
  %74 = load i32, i32* %73, align 4
  %75 = getelementptr %tuple, %tuple* %69, i32 0, i32 1
  store i32 %74, i32* %75, align 4
  %76 = getelementptr %list, %list* %const_list, i32 0, i32 0
  %77 = load i32, i32* %76, align 4
  %78 = getelementptr %list, %list* %const_list, i32 0, i32 1
  %79 = load i32, i32* %78, align 4
  %80 = getelementptr %list, %list* %l1, i32 0, i32 0
  %81 = getelementptr %list, %list* %l1, i32 0, i32 1
  store i32 %77, i32* %80, align 4
  store i32 %79, i32* %81, align 4
  %82 = getelementptr %list, %list* %const_list, i32 0, i32 2
  %83 = load %tuple.0*, %tuple.0** %82, align 8
  %84 = mul i32 32, %79
  %85 = call i8* (i32, ...) @_lfortran_malloc(i32 %84)
  %86 = bitcast i8* %85 to %tuple.0*
  %87 = getelementptr %list, %list* %l1, i32 0, i32 2
  store %tuple.0* %86, %tuple.0** %87, align 8
  %88 = alloca i32, align 4
  store i32 0, i32* %88, align 4
  br label %loop.head

loop.head:                                        ; preds = %loop.body, %.entry
  %89 = load i32, i32* %88, align 4
  %90 = icmp sgt i32 %77, %89
  br i1 %90, label %loop.body, label %loop.end

loop.body:                                        ; preds = %loop.head
  %91 = load i32, i32* %88, align 4
  %92 = getelementptr %list, %list* %const_list, i32 0, i32 2
  %93 = load %tuple.0*, %tuple.0** %92, align 8
  %94 = getelementptr inbounds %tuple.0, %tuple.0* %93, i32 %91
  %95 = getelementptr %list, %list* %l1, i32 0, i32 2
  %96 = load %tuple.0*, %tuple.0** %95, align 8
  %97 = getelementptr inbounds %tuple.0, %tuple.0* %96, i32 %91
  %98 = getelementptr %tuple.0, %tuple.0* %94, i32 0, i32 0
  %99 = load i32, i32* %98, align 4
  %100 = getelementptr %tuple.0, %tuple.0* %97, i32 0, i32 0
  store i32 %99, i32* %100, align 4
  %101 = getelementptr %tuple.0, %tuple.0* %94, i32 0, i32 1
  %102 = load double, double* %101, align 8
  %103 = getelementptr %tuple.0, %tuple.0* %97, i32 0, i32 1
  store double %102, double* %103, align 8
  %104 = getelementptr %tuple.0, %tuple.0* %94, i32 0, i32 2
  %105 = getelementptr %tuple.0, %tuple.0* %97, i32 0, i32 2
  %106 = getelementptr %tuple, %tuple* %104, i32 0, i32 0
  %107 = load double, double* %106, align 8
  %108 = getelementptr %tuple, %tuple* %105, i32 0, i32 0
  store double %107, double* %108, align 8
  %109 = getelementptr %tuple, %tuple* %104, i32 0, i32 1
  %110 = load i32, i32* %109, align 4
  %111 = getelementptr %tuple, %tuple* %105, i32 0, i32 1
  store i32 %110, i32* %111, align 4
  %112 = add i32 %91, 1
  store i32 %112, i32* %88, align 4
  br label %loop.head

loop.end:                                         ; preds = %loop.head
  br label %return

return:                                           ; preds = %loop.end
  ret void
}

declare i8* @_lfortran_malloc(i32, ...)

define i32 @main(i32 %0, i8** %1) {
.entry:
  call void @_lpython_set_argv(i32 %0, i8** %1)
  call void @__module__global_symbols__lpython_main_program()
  ret i32 0
}

declare void @_lpython_set_argv(i32, i8**)
code generation error: asr_to_llvm: module failed verification. Error:
Stored value type does not match pointer operand type!
  store %tuple* %const_tuple1, %tuple* %14, align 8
 %tuple = type { double, i32 }Stored value type does not match pointer operand type!
  store %tuple* %const_tuple3, %tuple* %36, align 8
 %tuple = type { double, i32 }Stored value type does not match pointer operand type!
  store %tuple* %const_tuple5, %tuple* %58, align 8
 %tuple = type { double, i32 }


Note: if any of the above error or warning messages are not clear or are lacking
context please report it to us (we consider that a bug that must be fixed).
```

### PR DESCRIPTION
**Pull request by kabra1110**
Fixes #1886.

### CODE REVIEW

**Comment by kabra1110** [comment]
Now, the issue seems to be [here](https://github.com/lcompilers/lpython/blob/27e8634b3d9b53632e3688574334e7be3e992957/src/libasr/codegen/asr_to_llvm.cpp#L1629).

Removing this line allows nested tuples
```
x: tuple[i32, tuple[i32, i32]] = (1, (2, 3))
```
but not
```
i: i32 = 0
y: tuple[i32, i32] = (i, i)
```

Keeping this as before allows the latter, but not nested tuples.

**Comment by kabra1110** [comment]
Fixed using separate `ptr_loads` value for structures.

**Comment by kabra1110** [comment]
Fixed using separate `ptr_loads` value for structures.

**Review by czgdp1807 — APPROVED** [review]

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 1]
```diff
@@ -2231,6 +2237,10 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
         ASR::Tuple_t* tuple_type_right = ASR::down_cast<ASR::Tuple_t>(ASRUtils::expr_type(x.m_right));
         std::string type_code_right = ASRUtils::get_type_code(tuple_type_right->m_type,
                                                         tuple_type_right->n_type);
+        ASR::Tuple_t* tuple_type = new ASR::Tuple_t();
```
Comment (certik) [inline src/libasr/codegen/asr_to_llvm.cpp:2240]: Don't initialize ASR types like this --- it leaks memory and is slow. Use the `ASR::make_Tuple_t` function.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 2]
```diff
@@ -2253,10 +2264,14 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
...
                                     is_list, m_dims, n_dims, a_kind));
+            tuple_type->m_type[tuple_type_left->n_type + i] = tuple_type_right->m_type[i];
         }
         llvm::Type* concat_tuple_type = tuple_api->get_tuple_type(type_code, llvm_el_types);
         llvm::Value* concat_tuple = builder->CreateAlloca(concat_tuple_type, nullptr, "concat_tuple");
-        tuple_api->concat(left, right, tuple_type_left, tuple_type_right, concat_tuple);
+        tuple_api->concat(left, right, tuple_type_left, tuple_type_right, concat_tuple,
+                          tuple_type, *module, name2memidx);
+        delete[] tuple_type->m_type;
+        delete tuple_type;
```
Comment (certik) [inline src/libasr/codegen/asr_to_llvm.cpp:2274]: These `delete` can be removed, once you use `make_Tuple_t`.

Also in general, never use raw `new` and `delete` in modern C++. Always use some kind of a smart pointer or allocator.

**Comment by certik** [comment]
@kabra1110 thanks for the PR!

Would you please mind fixing the issues from my comments above in a subsequent PR? If you need help let me know.

**Comment by kabra1110** [comment]
Sure, I will work on this.

**Comment by certik** [comment]
@kabra1110 awesome, thanks! If you can't figure out how to fix it, then please create an issue, so that we don't forget.

**Pull request merged**